### PR TITLE
auto dynamic shape

### PR DESCRIPTION
**Pull request by songbell**
### Details:
 - auto dynamic shape support

### Tickets:
 - https://jira.devtools.intel.com/browse/CVS-68362?filter=-1

### CODE REVIEW

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 1]
```diff
@@ -137,6 +137,39 @@ void MultiDeviceExecutableNetwork::GenerateWorkers(const std::string& device, co
     }
 }
 
+void MultiDeviceExecutableNetwork::IncreaseWorkers(AutoLoadContext& loadcontext, InferenceEngine::SoIInferRequestInternal& request_to_share) {
+    auto devicename = loadcontext.deviceInfo.deviceName;
+     auto& executableNetwork =  loadcontext.executableNetwork;
+    auto& idleWorkerRequests = _idleWorkerRequests[devicename];
+    //_inferPipelineTasksDeviceSpecific[devicename] = std::unique_ptr<ThreadSafeQueue<Task>>(new ThreadSafeQueue<Task>);
+    auto* idleWorkerRequestsPtr = &(idleWorkerRequests);
+    idleWorkerRequests.set_capacity(idleWorkerRequests.capacity() + 1);
```
Comment (tiger100256-hu) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:146]: need lock for  idleWorkerRequests.set_capacity(idleWorkerRequests.capacity() + 1);

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 2]
```diff
@@ -137,6 +137,42 @@ void MultiDeviceExecutableNetwork::GenerateWorkers(const std::string& device, co
...
+    auto devicename = loadcontext.deviceInfo.deviceName;
+     auto& executableNetwork =  loadcontext.executableNetwork;
+    auto& idleWorkerRequests = _idleWorkerRequests[devicename];
+    //_inferPipelineTasksDeviceSpecific[devicename] = std::unique_ptr<ThreadSafeQueue<Task>>(new ThreadSafeQueue<Task>);
+    auto* idleWorkerRequestsPtr = &(idleWorkerRequests);
+    {
+        std::lock_guard<std::mutex> lock(_idleWorkerMutex);
+        idleWorkerRequests.set_capacity(idleWorkerRequests.capacity() + 1);
+    }
+    auto* workerRequestPtr =  new WorkerInferRequest;
```
Comment (tiger100256-hu) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:150]: need destroy it in some place, if not , memory leak

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 3]
```diff
@@ -137,6 +137,42 @@ void MultiDeviceExecutableNetwork::GenerateWorkers(const std::string& device, co
...
+    auto& workerRequest = *workerRequestPtr;
+    workerRequest._inferRequest = { executableNetwork._so, executableNetwork->CreateInferRequest() };
+    IE_ASSERT(idleWorkerRequests.try_push(workerRequestPtr) == true);
```
Comment (tiger100256-hu) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:153]: IE_ASSERT(idleWorkerRequests.try_push(workerRequestPtr) == true); should  be after set callback, when workerRequest is totally ready

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 4]
```diff
@@ -486,9 +530,17 @@ InferenceEngine::IInferRequestInternal::Ptr MultiDeviceExecutableNetwork::Create
             auto& dev_requests = _workerRequests[_loadContext[ACTUALDEVICE].deviceInfo.deviceName];
             if (num < dev_requests.size()) {
                 request_to_share_blobs_with = dev_requests.at(num)._inferRequest;
+            } else {
+                IncreaseWorkers(_loadContext[ACTUALDEVICE], request_to_share_blobs_with);
+            }
+        } else {
+            auto& dev_requests = _workerRequests[_loadContext[CPU].deviceInfo.deviceName];
+            if (num < dev_requests.size()) {
```
Comment (tiger100256-hu) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:538]: num = _numRequestsCreated++; seem is not locked, I am not sure

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 2 above)
Comment (songbell) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:150]: yes, will try to figure out a way to handle this

**Comment by tiger100256-hu** [comment]
After customer destroy the Auto inferRequest, the real device inferRequest must is in the idleRequestQueue, 
when customer create new one, Auto will create new real device inferRequest, the old one can't be reused

**Comment by songbell** [comment]
> must



> After customer destroy the Auto inferRequest, the real device inferRequest must is in the idleRequestQueue, when customer create new one, Auto will create new real device inferRequest, the old one can't be reused

For auto, maybe it's acceptable if we bind auto infer request to actual infer request(without scheduling),  this will help resolve this issue because we will know when we can release the actual infer request.

**Comment by songbell** [comment]
> After customer destroy the Auto inferRequest, the real device inferRequest must is in the idleRequestQueue, when customer create new one, Auto will create new real device inferRequest, the old one can't be reused

updated PR, please help review again

**Comment by myshevts** [comment]
I was thinking that the change will be as simple as "if the network is dynamic the SelectDevice will return the CPU as the accelerator" leaving the rest of code completely intact... the suggested changes are vrey complex and fragile.

I wish we avoid touching the critical parts like scheduling, or on-the-fly changing the #requests, which is VERY fragile....
or.. let's discuss why we need all these on Mon sync... I can't get that from the code

**Comment by songbell** [comment]
> I was thinking that the change will be as simple as "if the network is dynamic the SelectDevice will return the CPU as the accelerator" leaving the rest of code completely intact... the suggested changes are vrey complex and fragile.
> 
> I wish we avoid touching the critical parts like scheduling, or on-the-fly changing the #requests, which is VERY fragile.... or.. let's discuss why we need all these on Mon sync... I can't get that from the code

sure, let's sync up on Mon. some thoughts from my side that the changes related with changing the #request and possibly the scheduling part are mainly from below 2 considerations:
1) from memory copy optimization perspective, like if used create more auto infer requests than the optimal numbers, AUTO:GPU and later AUTO:CPU,GPU case does not have enough hw memory to share
2) if user create more requests from like GPU backend, GPU  will create corresponding # requests, but for AUTO, currently the behavior is not like this, only the requests created right after the loadnetwork are available, maybe it's better to provide user the same experience like pure GPU backend in this case.

File: src/plugins/auto/executable_network.cpp [hunk 5]
```diff
@@ -503,7 +505,9 @@ InferenceEngine::IInferRequestInternal::Ptr MultiDeviceExecutableNetwork::Create
 
 InferenceEngine::IInferRequestInternal::Ptr MultiDeviceExecutableNetwork::CreateInferRequestImpl(InferenceEngine::InputsDataMap networkInputs,
                                                                                                 InferenceEngine::OutputsDataMap networkOutputs) {
+    _numRequestMutex.lock();
     auto num = _numRequestsCreated++;
+    _numRequestMutex.unlock();
```
Comment (myshevts) [inline src/plugins/auto/executable_network.cpp:510]: why we need the mutex? the _numRequestsCreated is already atomic

**Comment by myshevts** [comment]
Shall we have the logic so that when LoadNetwork gets the dynamic network as the input, only the CPU is selected as the device?

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 3 above)
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: not need this code now, close this review

File: src/plugins/auto/executable_network.cpp (hunk 5 above)
Comment (tiger100256-hu) [reply src/plugins/auto/executable_network.cpp:510]: when I reviewed this PR, I thought it's not safe for num here, I am not sure, it is like below code for me 
`lock();`
`mov eax _numRequestsCreated`
`unlock()`
`num = eax`
`lock()`
`_numRequestsCreated++`
`unlock()`

**Comment by wangleis** [comment]
> Shall we have the logic so that when LoadNetwork gets the dynamic network as the input, only the CPU is selected as the device?

Since VPU is supporting dynamic shapes at least for some networks, AUTO will not have the shortcut logic for dynamic network.

**Comment by songbell** [comment]
> Shall we have the logic so that when LoadNetwork gets the dynamic network as the input, only the CPU is selected as the device?

Hi Maxim,
I have previously tried below 2 methods but not working:
1) QueryNetwork, this does not work as expected, as discussed in another thread
2) I see there is an API in ngraph:
bool ov::Function::is_dynamic() const {
   ...
}
   But if using this API , for some network which have dynamic output, like *ssd-resnet34-1200-onnx*, this network can run on GPU, but the is_dynamic() will still report it as dynamic shape network , due to some output is dynamic(there is NMS layer in the network). In this case, it's not correct to report CPU as the target device.

I'm not sure if any other inferface/method that I might have missed...

File: src/plugins/auto/executable_network.cpp (hunk 5 above)
Comment (myshevts) [reply src/plugins/auto/executable_network.cpp:510]: I think this lock is completely redudnant (and would need to use lock_quard anyway) - and is not related to the subject of the PR

Comment (wangleis) [reply src/plugins/auto/executable_network.cpp]: Agree to separate this concern and follow up in next meeting. 

Let's merge this PR without this lock first. @myshevts

Comment (songbell) [reply src/plugins/auto/executable_network.cpp]: songbell Pending
operator ++ is atomic, but assign the ++ value to num may suffer the threading issue.
I have a simple test on this, and it will sporadically run into threading issues with the value of "num", easy to reproduce when thread num increases. 
remove it from this PR, and track if further if needed

**Comment by wangleis** [comment]
@ilya-lavrenov @dmitry-gorokhov 
Please help to review this PR which fix IE Core blob supporting dynamic shape network.

**Comment by myshevts** [comment]
> I'm not sure if any other inferface/method that I might have missed...
https://github.com/openvinotoolkit/openvino/pull/7788/files#diff-e7a536b4318b2270a4ee67f723b8ff57e9a5aed0928c11fa46e5a0b02db6300fR447

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/behavior/ov_infer_request/infer_request_dynamic.cpp [hunk 6]
```diff
@@ -18,6 +18,10 @@ const std::vector<std::map<std::string, std::string>> HeteroConfigs = {
     {{"TARGET_FALLBACK", CommonTestUtils::DEVICE_CPU}}
 };
 
+const std::vector<std::map<std::string, std::string>> AutoConfigs = {
+    {{InferenceEngine::MultiDeviceConfigParams::KEY_MULTI_DEVICE_PRIORITIES, CommonTestUtils::DEVICE_CPU}}
```
Comment (myshevts) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/behavior/ov_infer_request/infer_request_dynamic.cpp:22]: so we expclitly test the AUTO:CPU is working, but again, how about just "AUTO" - will it select the CPU (just because the GPU will fail loading the dynamic network)?
Shall we have such a test

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp [hunk 7]
```diff
@@ -468,7 +468,7 @@ void MKLDNNPlugin::MKLDNNInferRequest::SetBlob(const std::string& name, const In
                                << blobDesc.getPrecision() << ", if CNNNetwork output blob precision is: " << foundOutput->getPrecision();
         }
         size_t outputSize = foundOutput->getTensorDesc().getLayout() != InferenceEngine::Layout::SCALAR
-            ? InferenceEngine::details::product(foundOutput->getDims())
+            ? InferenceEngine::details::product(foundOutput->getTensorDesc().getDims())
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:471]: why is it needed?

File: src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp [hunk 8]
```diff
@@ -103,6 +103,8 @@ void IInferRequestInternal::SetBlob(const std::string& name, const Blob::Ptr& us
         !((foundInput && foundInput->getInputData()->isDynamic()) || (foundOutput && foundOutput->isDynamic()))) {
         IE_THROW() << "Input data is empty. Input name: \'" << name << "\'";
     }
+    const bool isInputDynamic = foundInput && foundInput->getInputData()->isDynamic();
```
Comment (ilya-lavrenov) [inline src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp:106]: this API is deprecated https://github.com/openvinotoolkit/openvino/blob/master/src/inference/include/ie/ie_data.h#L184-L185

BTW, why AUTO even need to call these methods? It should work with device-specific SetBlob / GetBlob like HETERO does.

See dynamism support in HETERO https://github.com/openvinotoolkit/openvino/pull/8012/files

CC @apankratovantonp

File: src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp [hunk 9]
```diff
@@ -164,14 +166,20 @@ Blob::Ptr IInferRequestInternal::GetBlob(const std::string& name) {
     DataPtr foundOutput;
     const SizeVector oneVector = {1};
     if (findInputAndOutputBlobByName(name, foundInput, foundOutput)) {
+        IE_SUPPRESS_DEPRECATED_START
+        const bool isInputDynamic = foundInput && foundInput->getInputData()->isDynamic();
+        IE_SUPPRESS_DEPRECATED_END
```
Comment (ilya-lavrenov) [inline src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp:171]: don't use deprecated API

File: src/plugins/auto/plugin.cpp [hunk 10]
```diff
@@ -595,4 +596,34 @@ std::vector<DeviceInformation> MultiDeviceInferencePlugin::FilterDevice(const st
...
+                        filterDevice.push_back(iter);
+                        return filterDevice;
+                    } else {
+                         IE_THROW(NotFound) << "No available device for dynamic shape network !";
+                    }
+                }
+            }
+        }
+        IE_SUPPRESS_DEPRECATED_END
+        return metaDevices;                                              }
```
Comment (ilya-lavrenov) [inline src/plugins/auto/plugin.cpp:628]: fix indentation

File: src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp (hunk 8 above)
Comment (songbell) [reply src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp:106]: Hi ilya, per my understanding, For hetero, each SetBlob/GetBlob of hetero infer request can be forwarded to sub-requests, which finally run on device-specific SetBlob/GetBlob. But for AUTO/Mutil, this is not the case. in auto/multi, the actual device infer request is been scheduled when doing stage1/stage2 of auto/multi async pipeline, which means when setting/getting blob, auto/multi does not know which actual device infer request will be scheduled for this auto/multi infer request. that's why I need to make change here to make auto work with dynamic shape.

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp (hunk 7 above)
Comment (songbell) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:471]: hi @ilya-lavrenov , in auto/multi plugin, we will need to call SetBlob/GetBlob for actual device infer request for both input/output buffers:
https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/auto/infer_request.cpp#L74

File: src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp (hunk 9 above)
Comment (songbell) [reply src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp:171]: @ilya, can you suggest which API I should use for dynamic case? I was referring to https://github.com/openvinotoolkit/openvino/blob/a77d6cfe36b49a90c9f49d35abe548dbfdd2e0d2/src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp#L103

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/behavior/ov_infer_request/infer_request_dynamic.cpp (hunk 6 above)
Comment (songbell) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/behavior/ov_infer_request/infer_request_dynamic.cpp:22]: add case for AUTO, AUTO:GPU,CPU and AUTO:GPU respectively

File: src/plugins/auto/plugin.cpp (hunk 10 above)
Comment (songbell) [reply src/plugins/auto/plugin.cpp]: fixed

**Comment by songbell** [comment]
> > I'm not sure if any other inferface/method that I might have missed...
> > https://github.com/openvinotoolkit/openvino/pull/7788/files#diff-e7a536b4318b2270a4ee67f723b8ff57e9a5aed0928c11fa46e5a0b02db6300fR447

add relevant logic following below:
https://github.com/openvinotoolkit/openvino/blob/b89889e98dd1065bd91322c44352769382bb8cdd/samples/cpp/benchmark_app/main.cpp#L447

File: src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp (hunk 9 above)
Comment (ilya-lavrenov) [reply src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp:171]: https://github.com/openvinotoolkit/openvino/blob/master/src/inference/dev_api/cpp_interfaces/interface/ie_iinfer_request_internal.hpp#L202-L203

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp (hunk 7 above)
Comment (ilya-lavrenov) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:471]: I suppose `foundOutput->getDims()` is the same `foundOutput->getTensorDesc().getDims()`, is not it?

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/skip_tests_config.cpp [hunk 11]
```diff
@@ -120,11 +120,13 @@ std::vector<std::string> disabledTestPatterns() {
         // Issue: CVS-66778
         R"(.*smoke_BehaviorTests.*InferFullyDynamicNetworkWith(S|G)etTensor.*)",
         R"(.*smoke_Hetero_BehaviorTests.*InferFullyDynamicNetworkWith(S|G)etTensor.*)",
+        R"(.*smoke_Auto_BehaviorTests.*InferFullyDynamicNetworkWith(S|G)etTensor.*)",
```
Comment (myshevts) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/skip_tests_config.cpp:123]: pls add comments why we have to skip some tests

File: src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp (hunk 8 above)
Comment (myshevts) [reply src/inference/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp:106]: right, Auto/Multi have to keep it's own blobs to be able to Set them to the actual device upon selection

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp (hunk 7 above)
Comment (songbell) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:471]: @ilya-lavrenov no, using foundOutput->getDims() will throw exception on below, this API does not support dynamic shape:

Expected: req.infer() doesn't throw an exception.
  Actual: it throws.

**/home/bell/work/mount/openvino/src/inference/src/ie_data.cpp:202 Cannot return dims for Data with dynamic shapes!**

File: inference-engine/tests/functional/plugin/gpu/subgraph_tests/shared_constant.cpp [hunk 12]
```diff
@@ -33,6 +33,9 @@ class SharedConstant : virtual public LayerTestsUtils::LayerTestsCommon {
         auto mul = std::make_shared<opset8::Multiply>(input4, constant);
         auto input5 = std::make_shared<opset8::Parameter>(type, constGroupConvBackpropShape);
         auto mul2 = std::make_shared<opset8::Multiply>(input5, constantGroupConv);
+        // explicitly set the output name, to avoid global conflict
+        mul2->set_friendly_name("Multiply_0");
```
Comment (myshevts) [inline inference-engine/tests/functional/plugin/gpu/subgraph_tests/shared_constant.cpp:37]: is this change really relevant to the PR?

**Review by myshevts — APPROVED** [review]

File: inference-engine/tests/functional/plugin/gpu/subgraph_tests/shared_constant.cpp (hunk 12 above)
Comment (songbell) [reply inference-engine/tests/functional/plugin/gpu/subgraph_tests/shared_constant.cpp:37]: actually this is a bug in this case, it will compare the outputs in the order of output map, not by retrieving the value from map key. with my change adding some test in GPU, it will trigger this bug, because the global node name for each ngraph node will change with more case added, thus will fail when comparing the outputs. I'm fixing it by setting the output names explicitly instead of letting the ie/ngraph generate it.

File: src/plugins/auto/plugin.cpp [hunk 13]
```diff
@@ -595,4 +596,24 @@ std::vector<DeviceInformation> MultiDeviceInferencePlugin::FilterDevice(const st
...
+                    filterDevice.push_back(iter);
+                    break;
+                }
+            }
+            if (filterDevice.size() == 0)
+                IE_THROW(NotFound) << "No available device for dynamic shape network !";
+            return filterDevice;
+        }
+    }
+    IE_SUPPRESS_DEPRECATED_END
```
Comment (ilya-lavrenov) [inline src/plugins/auto/plugin.cpp:616]: suggest to use ov::Model::is_dynamic here since OV 1.0 must not support dynamism.

Comment (ilya-lavrenov) [reply src/plugins/auto/plugin.cpp:616]: can be done in a separate PR

File: src/tests/functional/plugin/shared/src/behavior/ov_infer_request/infer_request_dynamic.cpp [hunk 14]
```diff
@@ -359,6 +359,20 @@ TEST_P(OVInferRequestDynamicTests, InferDynamicNetworkWithSetTensor2times) {
...
+    const ov::Shape refShape2 = inOutShapes[1].first;
+    const ov::Shape refOutShape = inOutShapes[0].second;
+    const ov::Shape refOutShape2 = inOutShapes[1].second;
+    std::map<std::string, ov::PartialShape> shapes;
+    shapes[tensor_name] = {ov::Dimension::dynamic(), 4, 20, 20};
+    OV_ASSERT_NO_THROW(function->reshape(shapes));
+    const std::string outputName = function->outputs().back().get_any_name();
+    // Load ov::Function to target plugins
+    ov::runtime::CompiledModel execNet;
+    ASSERT_THROW((execNet = ie->compile_model(function, targetDevice, configuration)), ov::Exception);
```
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/src/behavior/ov_infer_request/infer_request_dynamic.cpp:374]: why is it not supported?

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp (hunk 7 above)
Comment (ilya-lavrenov) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:471]: @maxnick why old API is used while dynamism can live internally?

File: src/tests/functional/plugin/shared/src/behavior/ov_infer_request/infer_request_dynamic.cpp (hunk 14 above)
Comment (songbell) [reply src/tests/functional/plugin/shared/src/behavior/ov_infer_request/infer_request_dynamic.cpp:374]: this case currently used by AUTO:GPU, the logic for AUTO is it will load network to CPU for dynamic network(AUTO, AUTO:CPU,GPU, AUTO:GPU,CPU will all load to CPU), and AUTO:GPU will not work in this case.

**Review by wangleis — APPROVED** [review]

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp (hunk 7 above)
Comment (maxnick) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:471]: @ilya-lavrenov , to answer your question, yes we have the old API that may be used in dynamic context, and the reason is that when we were enabling dynamic shapes support in the plugin, the new API was not ready. Obviously, we have to refactor the API implementation using the new paradigm to avoid such problems in the future.
Regarding the change itself, it seems that it is better to leave the `outputSize` computation as it was before, but move it under `if(!isDynamic)` guard check on L475 since the `outputSize` only makes sense under this check.

Comment (songbell) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp]: fixed, thanks for review @maxnick

File: src/plugins/auto/plugin.cpp (hunk 13 above)
Comment (songbell) [reply src/plugins/auto/plugin.cpp]: changed to ov::Model

**Review by maxnick — APPROVED** [review]
CPU part LGTM.

**Comment by ilya-lavrenov** [comment]
@songbell please, make CI green
